src/core/server/elasticsearch/integration_tests/client.test.ts

```diff
@@ -55,3 +63,67 @@ describe('elasticsearch clients', () => {
     expect(resp.headers!.warning).toMatch('system indices');
   });
 });
+
+function createFakeElasticsearchServer() {
+  const firstRequestReceived$ = new Subject<boolean>();
+
+  const requestHandler = jest.fn((req, res) => {
+    // Reply with a 200 and empty response by default (intentionally malformed response)
+    res.writeHead(200);
+    res.end();
+
+    firstRequestReceived$.next(true);
+  });
+
+  const server = http.createServer(requestHandler);
+  server.listen(esTestConfig.getPort());
+  return { server, requestHandler, firstRequestReceived$ };
+}
+
+describe('fake elasticsearch', () => {
+  let esServer: http.Server;
+  let requestHandler: jest.Mock;
+  let kibanaServer: Root;
+
+  beforeAll(async () => {
+    kibanaServer = createRootWithCorePlugins(
+      {
+        logging: { silent: false },
+        status: { allowAnonymous: true },
+      },
+      { verbose: true }
+    );
+    const fakeServer = await createFakeElasticsearchServer();
+    esServer = fakeServer.server;
+    requestHandler = fakeServer.requestHandler;
+
+    await kibanaServer.preboot();
+    await kibanaServer.setup();
+    kibanaServer.start();
+    await fakeServer.firstRequestReceived$.pipe(first()).toPromise(); // Waiting for the first ES request instead of the start to complete, because start can't never complete due to waiting for ES to be ready
```

Comment:
To be fair, AFAIK, once the client marks the connection as Product-invalid, it internally never checks again. It means that even when the ES server goes back to normal.

It might make sense to _crash_ Kibana because of that. Although, we might need a retry logic to handle any possible errors other than the product check.

As an additional thought: we don't stop when Kibana's credentials or the ES host config are wrong. 

I wonder if we could discuss this "Crash during startup on specific ES config errors" on a follow-up issue 🤔

Reply:
I just noticed about this sentence in the original issue: 
> Kibana should handle the incompatibility check exception thrown by @elastic/elasticsearch to log an error and stop Kibana gracefully.

I'll implement the shutdown logic 👍

Reply:
@afharo I added this requirement, but I didn't discuss it with the product management. Let's wait for @thesmallestduck feedback

Reply:
Agree with @mshustov , a shutdown with a log message seems preferable to me. It is important to give an indication of why Kibana isn't functioning. An indefinite hang without a log message when connecting to an incorrect ES version seems unpleasant.

Reply:
Deal! I'll trigger Kibana's shutdown when we see the ProductCheckError. FYI, we already log the error. But I'll make sure that it's more explicit.